packages/ics721/src/execute.rs

```diff
@@ -684,13 +721,35 @@ where
                 // Note, once cw721 doesn't support on-chain metadata yet - but this is where we will set
                 // that value on the debt-voucher token once it is supported.
                 // Also note that this is set for every token, regardless of if data is None.
-                TOKEN_METADATA.save(deps.storage, (class_id.clone(), id.clone()), &data)?;
+                IBC_RECEIVE_TOKEN_METADATA.save(
+                    deps.storage,
+                    (class_id.clone(), id.clone()),
+                    &data,
+                )?;
+
+                // parse token data and check whether it is of type NftExtension
+                let extension: Option<NftExtensionMsg> = match data {
+                    Some(data) => from_json::<NftExtension>(data)
+                        .ok()
+                        .map(|ext| NftExtensionMsg {
+                            animation_url: ext.animation_url,
+                            attributes: ext.attributes,
+                            background_color: ext.background_color,
+                            description: ext.description,
+                            external_url: ext.external_url,
+                            image: ext.image,
+                            image_data: ext.image_data,
+                            youtube_url: ext.youtube_url,
+                            name: ext.name,
+                        }),
+                    None => None,
+                };
```

Comment:
On receive, ics721 checks whether ibc packet holds nft extensions - if so NFT is minted with onchain metadata.